File: sync-project/src/gatherer/src/sync_kinects.cpp
#include <iostream>
using namespace std;

#include "ros/ros.h"
#include "sensor_msgs/Image.h"
#include "sensor_msgs/CameraInfo.h"
#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>


cv_bridge::CvImagePtr cv_ptr;
ros::Publisher kinect1_pub_image = n.advertise<Image>("kinect1_image", 1000);
ros::Publisher kinect2_pub_image = n.advertise<Image>("kinect2_image", 1000);
ros::Publisher kinect3_pub_image = n.advertise<Image>("kinect3_image", 1000);

ros::Publisher kinect1_pub_depth = n.advertise<Image>("kinect1_depth", 1000);
ros::Publisher kinect2_pub_depth = n.advertise<Image>("kinect2_depth", 1000);
ros::Publisher kinect3_pub_depth = n.advertise<Image>("kinect3_depth", 1000);

void callback(const ImageConstPtr& kinect1,const ImageConstPtr& kinect2,const ImageConstPtr& kinect3){
    kinect1_pub_image.publish(kinect1)
    kinect2_pub_image.publish(kinect2)
    kinect3_pub_image.publish(kinect3)
}

void callback_depth(const ImageConstPtr& kinect1,const ImageConstPtr& kinect2,const ImageConstPtr& kinect3){
    kinect1_pub_depth.publish(kinect1)
    kinect2_pub_depth.publish(kinect2)
    kinect3_pub_depth.publish(kinect3)
}

void save_image(const ImageConstPtr& image,const String& path){
    
    try{
        cv_ptr = cv_bridge::toCvCpy(msg, sensor_msgs::image_encodings::BGR8):

    }catch(cv_bridge::Exception& e){
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
    }

    cv::imwrite(path, cv_ptr->image);
}

int main(int argc, char **argv){
    ros::init(argc,argv, "Gatherer");
    ros::NodeHandle n("~");
    //ros::Subscriber sub = n.subscribe("/kinect1/rgb/image_raw")
    message_filters::Subscriber<Image> kinect1(n, "/kinect1/rgb/image",1);
    message_filters::Subscriber<Image> kinect2(n, "/kinect2/rgb/image",1);
    message_filters::Subscriber<Image> kinect3(n, "/kinect3/rgb/image",1);

    TimeSynchronizer<Image,Image,Image> sync(kinect1,kinect2,kinect3,1000);
    sync.registerCallback(boost::bind(&callback,_1,_2,_3));

    message_filters::Subscriber<Image> kinect1_d(n, "/kinect1/depth/image",1);
    message_filters::Subscriber<Image> kinect2_d(n, "/kinect2/depth/image",1);
    message_filters::Subscriber<Image> kinect3_d(n, "/kinect3/depth/image",1);

    /*
    typedef sync_policies::ApproximateTime<Image, Image, Image> MySyncPolicy;
    // ApproximateTime takes a queue size as its constructor argument, hence MySyncPolicy(10)
    Synchronizer<MySyncPolicy> sync(MySyncPolicy(10), kinect1, kinect2,kinect3);
    */
    TimeSynchronizer<Image,Image,Image> sync(kinect1,kinect2,kinect3,1000);
    sync.registerCallback(boost::bind(&callback_depth,_1,_2,_3));
    ros::spin();

    return 0;
}

File: sync_kinects.py
import rospy
import cv2 as cv
import sys
import os
from std_msgs.msg import String
from sensor_msgs.msg import Image
from geometry_msgs.msg import Twist
from cv_bridge import CvBridge, CvBridgeError
import message_filters

PATH = os.path.abspath('images')
PATH = PATH + "/images"
print('Saving path = {}'.format(PATH))

class Gatherer:
    def __init__(self):
        self.frame_image = 0
        self.frame_depth = 0
        self.bridge = CvBridge()
        self.kinect1 = message_filters.Subscriber('/kinect1/rgb/image_raw', Image)
        self.kinect2 = message_filters.Subscriber('/kinect2/rgb/image_raw', Image)
        self.kinect3 = message_filters.Subscriber('/kinect3/rgb/image_raw', Image)
        self.timeSync = message_filters.ApproximateTimeSynchronizer([self.kinect1, self.kinect2, self.kinect3 ], queue_size=10,slop=0.1, allow_headerless=True)
        self.timeSync.registerCallback(self.SyncCallbackImage)

        self.kinect1_d = message_filters.Subscriber('/kinect1/depth/image_raw', Image)
        self.kinect2_d = message_filters.Subscriber('/kinect2/depth/image_raw', Image)
        self.kinect3_d = message_filters.Subscriber('/kinect3/depth/image_raw', Image)
        self.timeSync = message_filters.ApproximateTimeSynchronizer([self.kinect1_d, self.kinect2_d, self.kinect3_d ], queue_size=10,slop=0.1, allow_headerless=True)
        self.timeSync.registerCallback(self.SyncCallbackDepth)
        

    def SyncCallbackImage(self, kinect1, kinect2, kinect3):
        self.save_image(kinect1, "/kinect1/image")
        self.save_image(kinect2, "/kinect2/image")
        self.save_image(kinect2, "/kinect3/image")
        self.frame_image +=1
    def SyncCallbackDepth(self, kinect1, kinect2, kinect3):
        self.save_depth(kinect1, "/kinect1/depth")
        self.save_depth(kinect2, "/kinect2/depth")
        self.save_depth(kinect2, "/kinect3/depth")
        self.frame_depth +=1
        
    def save_image(self, img, path):
        cv_image = self.bridge.imgmsg_to_cv2(img, "bgr8")
        cv.imwrite(PATH+path + "_" + str(self.frame_image) + ".jpg", cv_image)  
    def save_depth(self, img, path):
        cv_image = self.bridge.imgmsg_to_cv2(img, "bgr8")
        cv.imwrite(PATH+path + "_" + str(self.frame_depth) + ".jpg", cv_image) 


def main():
    handler = Gatherer()
    rospy.init_node('Gatherer', anonymous=True)
    rate = rospy.Rate(2)
    try:
        rospy.spin()
    except KeyboardInterrupt:
        print("Shutting down")

if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(e)